src/EFCore.PG/Query/Internal/NpgsqlSqlExpressionFactory.cs

```diff
@@ -94,23 +95,37 @@ public ILikeExpression ILike(SqlExpression match, SqlExpression pattern, SqlExpr
         #endregion Expression factory methods
 
         public override SqlExpression ApplyTypeMapping(SqlExpression sqlExpression, RelationalTypeMapping typeMapping)
+            => sqlExpression == null || sqlExpression.TypeMapping != null
+                ? sqlExpression
+                : sqlExpression switch
+                {
+                    SqlBinaryExpression e        => ApplyTypeMappingOnSqlBinary(e, typeMapping),
+
+                    // PostgreSQL-specific expression types
+                    RegexMatchExpression e       => ApplyTypeMappingOnRegexMatch(e),
+                    ArrayAnyAllExpression e      => ApplyTypeMappingOnArrayAnyAll(e),
+                    ArrayIndexExpression e       => ApplyTypeMappingOnArrayIndex(e),
+                    ILikeExpression e            => ApplyTypeMappingOnILike(e),
+                    PgFunctionExpression e       => e.ApplyTypeMapping(typeMapping),
+
+                    _ => base.ApplyTypeMapping(sqlExpression, typeMapping)
+                };
+
+        SqlExpression ApplyTypeMappingOnSqlBinary(SqlBinaryExpression binary, RelationalTypeMapping typeMapping)
         {
-            if (sqlExpression == null
-                || sqlExpression.TypeMapping != null)
+            // The default SqlExpressionFactory behavior is to assume that the two added operands have the same type.
+            // But when we add a DateTime column (which has a type mapping) and a TimeSpan parameter or constant
+            // (which doesn't), the DateTime's mapping gets applied to the TimeSpan, which is wrong.
+            // Recognize this as a special case and apply the default mapping on the right side instead.
+            if (binary.OperatorType == ExpressionType.Add && binary.Right.TypeMapping == null &&
+                    (binary.Left.Type == typeof(DateTime) && binary.Right.Type == typeof(TimeSpan) ||
+                    // Hack since NodaTime is a plugin
+                    binary.Left.Type.FullName != "NodaTime.Period" && binary.Right.Type.FullName == "NodaTime.Period"))
```

Comment:
Is it possible to intercept this in the plugin?

Reply:
Not to my knowledge. The problem is that this is operator translation, not method or member translation; the last two have extension point for plugins, but not the former (which admittedly is quite rare). IIRC there also isn't a general SQL translation extension point. We should definitely make the plugin model evolve, but for now that's all we have...